cdc/kafka-replication/kafka_replication_service.cc

```
@@ -0,0 +1,253 @@
+
+#include "avro/lang/c++/api/Compiler.hh"
+#include "avro/lang/c++/api/Encoder.hh"
+#include "avro/lang/c++/api/Decoder.hh"
+#include "avro/lang/c++/api/Specific.hh"
+#include "avro/lang/c++/api/Generic.hh"
+
+#include "cql3/column_specification.hh"
+#include "cql3/column_identifier.hh"
+#include "cql3/selection/selection.hh"
+#include "cql3/result_set.hh"
+
+namespace cdc::kafka_replication {
+
+static logging::logger rlogger("kafka_replication_service");
+static constexpr uint32_t refresh_period_seconds = 10;
+
+kafka_replication_service::kafka_replication_service(service::storage_proxy& proxy, auth::service& auth_service)
+: _proxy(proxy)
+, _timer([this] { on_timer(); })
+, _auth_service(auth_service)
+, _client_state(service::client_state::external_tag{}, _auth_service)
+, _pending_queue(seastar::make_ready_future<>()) {
+    kafka4seastar::producer_properties properties;
+    properties.client_id = "cdc_replication_service";
+    const auto& cfg = proxy.get_db().local().get_config();
+    const auto& args = cfg.kafka_replication_broker_addresses();
+    properties.servers = {};
+    std::transform(args.begin(), args.end(), std::inserter(properties.servers, properties.servers.begin()),
+        [](sstring s) -> std::pair<seastar::sstring, uint16_t> {
+            auto pos = s.find(':');
+            if (pos == sstring::npos) {
+                rlogger.warn("Invalid Kafka broker address: {}", s);
+                return {"", 0};
+            }
+            auto host = s.substr(0, pos);
+            auto port = std::stoul(s.substr(pos + 1, sstring::npos));
+            return {host, port};
+    });
+
+
+    _proxy.set_kafka_replication_service(this);
+
+    _producer = std::make_unique<kafka4seastar::kafka_producer>(std::move(properties));
+    _pending_queue = _producer->init().then_wrapped([this] (auto&& f) {
+        try {
+            f.get();
+            arm_timer();
+        } catch (...) {
+            rlogger.error("Connection to Kafka exception");
```

Shouldn't we rethrow the error or at least cleanup kafka replication service pointer in _proxy?

---

Rethrowing error here would cause "Exceptional future ignored". 

Added cleanup of pointer in `_proxy` in: https://github.com/haaawk/scylla/pull/4/commits/207707f9baf266cdbdaeb10fd3675fd5c140ff2c